### PR TITLE
Added a way to expire binaries and payloads

### PR DESCRIPTION
**Pull request by joseph-giron**
This was wanted in the wish-list. 

The change adds an exe option 'expire' which takes the current month / year and embeds them into the payload generation code. If the month and year match, then the payload executes like normal (virtualalloc RWE, copy shellcode, run), otherwise, the program just exits gracefully. 

## Verification

List the steps needed to make sure this thing works

- [ ] Start `msfconsole`
- [ ] `use payload/windows/whatever
- [ ] generate -t exe -f whatever.exe
Assuming you call the win32_rwx_exec_expire(code) function, it should generate the assembly code directly into the payload file with the expiration date of the current month / year.
doesitwork
msf_expire
launch

### CODE REVIEW

**Review by sempervictus — COMMENTED** [review]
Awesome, Framework needed this...
Will add to test queue.
Any chance you've got a 64bit version cooking somewhere? :)

Ping @busterb - could you confirm its OK to add this functionality in Framework at the Exe level?

**Comment by joseph-giron** [comment]
Only win32 has code in the exe.rb file. A linux or win64 variant would require a generated template file embedding the current date.  Do you know off hand if template executables are assembled / compiled by MSF as they're needed(on the fly) or if payloads just cram code into them at particular offsets? If the first is correct, then a ruby file that echos the assembly code to create the win64 exe or linux or what have you would be easy to add.

**Comment by sempervictus** [comment]
MSF can inject templates, diff bins with diversion to shellcode, and assemble full PEs from scratch. Util::Exe is one of the things which will need some cleanup by us all down the line, but a ton of the generation is done in core/payload. Take a look at how metasm payloads get assembled. For instance, on win32 rev_TCP got sliced up so we could shim other block_recv sections. So having this check preempt stage execution/sock ops keeps all questionable activity from occurring.

File: lib/msf/util/exe.rb [hunk 1]
```diff
@@ -237,9 +238,17 @@ def self.to_win32pe(framework, code, opts = {})
 
     # Allow the user to specify their own EXE template
     set_template_default(opts, "template_x86_windows.exe")
-
+	
+	# Finally got this working
```
Comment (jmartin-r7) [inline lib/msf/util/exe.rb]: Use space for indentation please. Also some comments look like WIP please remove those are no longer needed.

File: lib/msf/core/exploit/exe.rb [hunk 2]
```diff
@@ -17,6 +17,7 @@ def initialize(info = {})
     register_advanced_options(
       [
         OptBool.new('EXE::EICAR',    [false, 'Generate an EICAR file instead of regular payload exe']),
+		OptBool.new('EXE::Expire',   [false, 'Expire payload after 1 month']),
```
Comment (jmartin-r7) [inline lib/msf/core/exploit/exe.rb]: Use space for indentation please.

File: lib/msf/util/exe.rb [hunk 3]
```diff
@@ -1828,7 +2069,8 @@ def self.win32_rwx_exec(code)
     enc = Metasm::Shellcode.assemble(Metasm::Ia32.new, wrapper).encoded
     enc.data + code
   end
-
+  
+
```
Comment (acammack-r7) [inline lib/msf/util/exe.rb]: Also, please remove the trailing spaces on these lines.

File: lib/msf/core/exploit/exe.rb (hunk 2 above)
Comment (sempervictus) [reply lib/msf/core/exploit/exe.rb]: Maybe make this a configurable option? ExpireTime ?

Comment (bwatters-r7) [reply lib/msf/core/exploit/exe.rb]: Can we kick this up to a higher level so it is not limited to exe?

**Comment by bwatters-r7** [comment]
Is there any movement on this?

**Comment by joseph-giron** [comment]
Thanks for reminding me. I'll take a peek at it this weekend see if we
can't get it out the door soon.

**Comment by acammack-r7** [comment]
Hello @joseph-giron! To help us have a better grasp on which and how many PRs are in active development, I am closing this for now. Having expiration built into payloads is definitely a thing we want, though, so I am adding the `attic` label so we don't lose track of this. You can learn about our new PR label [on our wiki](https://github.com/rapid7/metasploit-framework/wiki/So-Your-PR-was-closed#moved-to-the-attic). Feel free to reopen once you pick this back up. I think replacing the hard tabs with two spaces each, having a configurable expiration time, and removing WIP comments and trailing spaces would get this well on its way out the door.

**Pull request closed without merging**